[#65747] Make /scim_v2/Schemas endpoint return required by OP schemas.

ba1ash opened the pull request:
https://community.openproject.org/work_packages/65747

- Make /scim_v2/Schemas endpoint return OP required schemas. Set required field to be required(givenName, familyName, email etc.). Set caseSensetive to be true where it is case sensetive. Remove fields that are not used by OpenProject. Note: making fields as requried make scimitar check their presence and responds with 400 which is useful.
- Refactor SCIM API controllers by extracting error handling in a general(🫡) method
- Add extra specs
- Restyle SCIM API specs a bit with hope that it makes them more readable...

ba1ash committed:
[#65747] Make /scim_v2/Schemas endpoint return required by OP schemas.

https://community.openproject.org/work_packages/65747

- Make /scim_v2/Schemas endpoint return OP required schemas.
  Set required field to be required(givenName, familyName, email etc.).
  Set caseSensetive to be true where it is case sensetive.
  Remove fields that are not used by OpenProject.
  Note: making fields as requried make scimitar check their presence and
  responds with 400 which is useful.
- Refactor SCIM API controllers by extracting error handling in a general(🫡) method
- Add extra specs
- Restyle SCIM API specs a bit with hope that it makes them more readable...

NobodysNightmare commented on app/controllers/scim_v2/base_controller_actions.rb:
```diff
@@ -89,6 +89,27 @@ def include_attributes
+      def raise_result_errors_for_scim(result)
+        result.on_failure do |result|
+          uniqueness_error = result.errors.find { |e| e.type == :taken }
+          unauthorized_error = result.errors.find { |e| e.type == :error_unauthorized }
```
:green_circle: We could extract these two into helper methods: `uniqueness_error?(result)` and `authorization_error?(result)`

This would help Rubocop accept this method, because those helpers could easily be used in the ifs below directly.

Also I'd suggest using `any?` instead of `find`, because we are only interested in the existence of those errors, not their actual values.

NobodysNightmare commented on lib/scimitar_schema_extension.rb:
```diff
@@ -28,11 +28,43 @@
+  def self.scim_attributes
+    @scim_attributes ||= [
+      Scimitar::Schema::Attribute.new(name: "familyName", caseExact: true, type: "string", required: true),
+      Scimitar::Schema::Attribute.new(name: "givenName",  caseExact: true, type: "string", required: true)
```
:yellow_circle: I am a bit irritated by the SCIM specs wording on what `caseExact` means. On the one hand it says "The server SHALL use case sensitivity when evaluating filters", which I would deny us doing for names. I'd expect us to find `Jane Doe` even if you search for `jane doe`, wouldn't I?

On the other hand they write "For attributes that are case exact, the server SHALL preserve case for any value submitted", which we will definitely do for names.

Have you compared this to what Scimitar would have returned by default? What's your take on the meaning of `caseExact`?

Full text of the spec:
```
caseExact  A Boolean value that specifies whether or not a string
         attribute is case sensitive.  The server SHALL use case
         sensitivity when evaluating filters.  For attributes that are
         case exact, the server SHALL preserve case for any value
         submitted.  If the attribute is case insensitive, the server
         MAY alter case for a submitted value.  Case sensitivity also
         impacts how attribute values MAY be compared against filter
         values (see [Section 3.4.2.2 of [RFC7644]](https://www.rfc-editor.org/rfc/rfc7644#section-3.4.2.2)).
```

ba1ash replied:
I checked actual behavior one more time and walked though scimitar docs and source code. So, this is what I found.
Actually having `caseExact` set in schema does not influence query construction. Only `id`, `externalId` and `meta.*` are matched case-sensitively. 
Here there are a few paragraphs on the topic:[https://github.com/pond/scimitar/blob/09b284778340ee067df9e5140d0230386afb9992/README.md#omissions
Therefore, I adjusted schemas returned by `/Schemas` endpoint.

NobodysNightmare commented on lib/scimitar_schema_extension.rb:
```diff
@@ -28,11 +28,43 @@
+    def scim_attributes
+      [
+        Scimitar::Schema::Attribute.new(name: "userName", caseExact: true, type: "string", uniqueness: "server", required: true),
+        Scimitar::Schema::Attribute.new(name: "name", caseExact: true, complexType: OpenProjectNameComplexType, required: true),
```
is `caseExact` required here, since this is a complex type that consists of pieces that are case exact or not. I understand that `required` is needed both here and in the complex type, because we are neither allowed to skip the complex type nor its members.

ba1ash replied:
Yes, you understood correctly. Let's have it in both places.

NobodysNightmare commented on spec/requests/scim_v2/schemas_spec.rb:
```diff
@@ -42,26 +42,93 @@
+          "meta" => { "resourceType" => "Schema",
+                      "location" => "http://test.host/scim_v2/Schemas?name=urn%3Aietf%3Aparams%3Ascim%3Aschemas%3Acore%3A2.0%3AGroup" },
+          "attributes" => [
+            { "multiValued" => false, "required" => true, "caseExact" => true, "mutability" => "readWrite", "uniqueness" => "none", "returned" => "default", "name" => "displayName", "type" => "string" },
```
I see that we disabled the cop for line length before, but I believe that was done to allow writing the URL above in a single line. Since this is a hash, I'd say we should still write this in multiple lines for better readability. (the example has "too many lines" anyways, so we can exceed that even further xD)

ba1ash replied:
I changed formatting.

NobodysNightmare reviewed (changes requested):
I have one question where the SCIM specification confuses me. Let's be sure we are aligned on how we want to handle that.

The other remarks are mostly styling-related.

ba1ash committed:
[#65747] Make SCIM /Schemas reflect real behavior.

- Refactor SCIM error handling logic.
- Make SCIM /Schemas reflect real behavior.
  Add `id` attribute.
  Change caseExact to be false everywhere excerpt `id` and `externalId`.
  See https://github.com/pond/scimitar/blob/09b284778340ee067df9e5140d0230386afb9992/README.md#omissions
  for details. And I checked actual source code. There is no place where caseExact values specified in schema
  are respected. Only `id`, `externalId` and `meta.*` are matched case-sensitively.
- Extend filtering specs with upcased/downcased search terms.
- Change formatting of long spec examples.
